drivers/stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,349 @@
+{
+	stepper_motor_set_coil_charge(dev);
+	update_coil_charge(dev);
+	update_remaining_steps(dev->data);
+}
+
+static void velocity_mode_task(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	stepper_motor_set_coil_charge(dev);
+	update_coil_charge(dev);
+	k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
+}
+
+static void stepper_work_step_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct gpio_stepper_motor_controller_data *data =
+		CONTAINER_OF(dwork, struct gpio_stepper_motor_controller_data, stepper_dwork);
+
+	K_SPINLOCK(&data->lock) {
+		switch (data->run_mode) {
+		case STEPPER_POSITION_MODE:
+			position_mode_task(data->dev);
+			break;
+		case STEPPER_VELOCITY_MODE:
+			velocity_mode_task(data->dev);
+			break;
+		default:
+			LOG_WRN("Unsupported run mode %d", data->run_mode);
+			break;
+		}
+	}
+}
+
+static int gpio_stepper_move(const struct device *dev, int32_t micro_steps,
+			     struct k_poll_signal *async)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	if (data->delay_in_us == 0) {
+		LOG_ERR("Velocity not set or invalid velocity set");
+		return -EINVAL;
+	}
+	K_SPINLOCK(&data->lock) {
+		if (data->async_signal) {
+			k_poll_signal_reset(data->async_signal);
+		}
+		data->async_signal = async;
+		data->run_mode = STEPPER_POSITION_MODE;
+		data->step_count = micro_steps;
+		update_direction_from_step_count(data);
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
+
+static int gpio_stepper_set_actual_position(const struct device *dev, int32_t position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->actual_position = position;
+	}
+	return 0;
+}
+
+static int gpio_stepper_get_actual_position(const struct device *dev, int32_t *position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		*position = data->actual_position;
+	}
+	return 0;
+}
+
+static int gpio_stepper_set_target_position(const struct device *dev, int32_t position,
+					    struct k_poll_signal *async)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	if (data->delay_in_us == 0) {
+		LOG_ERR("Velocity not set or invalid velocity set");
+		return -EINVAL;
+	}
+	K_SPINLOCK(&data->lock) {
+		if (data->async_signal) {
+			k_poll_signal_reset(data->async_signal);
+		}
+		data->async_signal = async;
+		data->run_mode = STEPPER_POSITION_MODE;
+		data->step_count = position - data->actual_position;
+		update_direction_from_step_count(data);
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
+
+static int gpio_stepper_is_moving(const struct device *dev, bool *is_moving)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	*is_moving = k_work_delayable_is_pending(&data->stepper_dwork);
+	LOG_DBG("Is Motor Moving %d", *is_moving);
```

Comment:
Maybe ternary a string to get a more friendly message from debug instead of a bool
```suggestion
	LOG_DBG("Motor is %s moving", *is_moving ? "" : "not" );
```